// @supabase/storage-js/src/lib/vectors/index.ts
// Main client
export { StorageVectorsClient, VectorBucketScope, VectorIndexScope } from '@supabase/storage-js/src/lib/vectors/StorageVectorsClient'
export type { StorageVectorsClientOptions } from '@supabase/storage-js/src/lib/vectors/StorageVectorsClient'

// API classes (for advanced usage)
export { default as VectorBucketApi } from '@supabase/storage-js/src/lib/vectors/VectorBucketApi'
export { default as VectorIndexApi } from '@supabase/storage-js/src/lib/vectors/VectorIndexApi'
export { default as VectorDataApi } from '@supabase/storage-js/src/lib/vectors/VectorDataApi'
export type { CreateIndexOptions } from '@supabase/storage-js/src/lib/vectors/VectorIndexApi'

// Types
export type {
  // Core types
  VectorBucket,
  VectorIndex,
  VectorData,
  VectorMetadata,
  VectorObject,
  VectorMatch,
  EncryptionConfiguration,
  MetadataConfiguration,
  VectorDataType,
  DistanceMetric,
  VectorFilter,

  // Request/Response types
  ListVectorBucketsOptions,
  ListVectorBucketsResponse,
  ListIndexesOptions,
  ListIndexesResponse,
  PutVectorsOptions,
  GetVectorsOptions,
  GetVectorsResponse,
  DeleteVectorsOptions,
  ListVectorsOptions,
  ListVectorsResponse,
  QueryVectorsOptions,
  QueryVectorsResponse,

  // Response wrappers
  ApiResponse,
  SuccessResponse,
  ErrorResponse,
  VectorFetchParameters,
} from '@supabase/storage-js/src/lib/vectors/types'

// Errors
export {
  StorageVectorsError,
  StorageVectorsApiError,
  StorageVectorsUnknownError,
  StorageVectorsErrorCode,
  isStorageVectorsError,
} from '@supabase/storage-js/src/lib/vectors/errors'

// Fetch utilities (for custom implementations)
export type { Fetch, FetchOptions, RequestMethodType } from '@supabase/storage-js/src/lib/vectors/fetch'

// Helper utilities
export {
  resolveFetch,
  resolveResponse,
  isPlainObject,
  normalizeToFloat32,
  validateVectorDimension,
} from '@supabase/storage-js/src/lib/vectors/helpers'
